File: frontend/js/utils/permissions.js
function hasPermission(user, key) {
  if (!user) return false;
  if (user.role === "admin") return true;
  const perms = user.permissions || {};
  return Boolean(perms[key]);
}

function hasMenuPermission(user, key) {
  if (!user) return false;
  if (user.role === "admin") return true;
  const perms = user.permissions || {};
  return Boolean(perms[key]);
}

export function canManageUsers(user) {
  return user?.role === "admin";
}

export function canViewPropusks(user) {
  return hasPermission(user, "view");
}

export function canCreatePropusks(user) {
  return hasPermission(user, "create");
}

export function canEditPropusks(user) {
  return hasPermission(user, "edit");
}

export function canActivatePropusks(user) {
  return hasPermission(user, "activate");
}

export function canDeletePropusks(user) {
  return hasPermission(user, "delete");
}

export function canAnnulPropusks(user) {
  return hasPermission(user, "annul");
}

export function canMarkDelete(user) {
  return hasPermission(user, "mark_delete");
}

export function canEditOrganizations(user) {
  return hasPermission(user, "edit_organization");
}

export function canDownload(user) {
  return hasPermission(user, "download_pdf");
}

export function canShowMenuPropusks(user) {
  return hasMenuPermission(user, "menu_propusks") && hasPermission(user, "view");
}

export function canShowMenuHome(user) {
  return hasMenuPermission(user, "menu_home");
}

export function canShowMenuReferences(user) {
  return hasMenuPermission(user, "menu_references");
}

export function canShowMenuPrint(user) {
  return hasMenuPermission(user, "menu_print");
}

export function canShowMenuReports(user) {
  return hasMenuPermission(user, "menu_reports");
}

export function canShowMenuUsers(user) {
  return hasMenuPermission(user, "menu_users");
}

export function canShowMenuSettings(user) {
  return hasMenuPermission(user, "menu_settings");
}

File: auth/permissions.py
PERMISSION_KEYS = [
    "view",
    "create",
    "edit",
    "delete",
    "annul",
    "mark_delete",
    "activate",
    "edit_organization",
    "download_pdf",
    "menu_propusks",
    "menu_home",
    "menu_references",
    "menu_print",
    "menu_reports",
    "menu_users",
    "menu_settings",
    "menu_create_pass",
    "menu_edit_pass",
    "menu_history",
]


def normalize_permissions(payload):
    if payload is None:
        return None
    if isinstance(payload, dict):
        data = {key: bool(payload.get(key, False)) for key in PERMISSION_KEYS}
        return data
    if isinstance(payload, (list, tuple, set)):
        data = {key: False for key in PERMISSION_KEYS}
        for item in payload:
            if item in data:
                data[item] = True
        return data
    return {key: False for key in PERMISSION_KEYS}


ROLE_DEFAULTS = {
    "admin": {key: True for key in PERMISSION_KEYS},
    "manager": {
        "view": True,
        "create": True,
        "edit": True,
        "delete": False,
        "annul": True,
        "mark_delete": True,
        "activate": True,
        "edit_organization": False,
        "download_pdf": True,
        "menu_home": True,
        "menu_propusks": True,
        "menu_references": True,
        "menu_print": True,
        "menu_reports": True,
        "menu_users": False,
        "menu_settings": False,
        "menu_create_pass": False,
        "menu_edit_pass": False,
        "menu_history": False,
    },
    "guard": {
        "view": True,
        "create": False,
        "edit": False,
        "delete": False,
        "annul": False,
        "mark_delete": False,
        "activate": False,
        "edit_organization": False,
        "download_pdf": False,
        "menu_home": True,
        "menu_propusks": True,
        "menu_references": False,
        "menu_print": False,
        "menu_reports": False,
        "menu_users": False,
        "menu_settings": False,
        "menu_create_pass": False,
        "menu_edit_pass": False,
        "menu_history": False,
    },
    "viewer": {
        "view": True,
        "create": False,
        "edit": False,
        "delete": False,
        "annul": False,
        "mark_delete": False,
        "activate": False,
        "edit_organization": False,
        "download_pdf": False,
        "menu_home": True,
        "menu_propusks": True,
        "menu_references": False,
        "menu_print": False,
        "menu_reports": False,
        "menu_users": False,
        "menu_settings": False,
        "menu_create_pass": False,
        "menu_edit_pass": False,
        "menu_history": False,
    },
}


def defaults_for_role(role):
    return ROLE_DEFAULTS.get(role, ROLE_DEFAULTS["viewer"]).copy()
